File: java/simple-cipher/src/main/java/Cipher.java
import java.util.Random;
import java.util.function.IntPredicate;
import java.util.stream.IntStream;

public class Cipher {
    private final String key;

    private static final IntPredicate tureP = (i)-> true;

    public Cipher(String key) {
        if(key.isEmpty() ||
                key.chars().anyMatch((i) -> !Character.isLowerCase(i))) {
            throw new IllegalArgumentException("Incorrect Key: " + key);
        }

        this.key = key;
    }

    public Cipher() {
        this.key = randomString(100);
    }


    public static String randomString(int length) {
        final Random rand = new Random();
        return rand.ints('a', 'z' + 1)
                .limit(length)
                .collect(StringBuilder::new, StringBuilder::appendCodePoint, StringBuilder::append)
                    .toString();
    }

    public String getKey() {
        return key;
    }


    public static int normalize(int ch) {
        return ('a' <= ch && ch <= 'z')? ch:
                (ch < 'a')? normalize(ch + 26): normalize(ch - 26);
    }

    public String encode(String plain) {
        return IntStream.range(0, plain.length())
                .map(i-> plain.charAt(i) + (key.charAt(i) - 'a'))
                .map(Cipher::normalize)
                .collect(StringBuilder::new, StringBuilder::appendCodePoint, StringBuilder::append)
                    .toString();
    }

    public String decode(String cipher) {
        return IntStream.range(0, cipher.length())
                .map(i-> cipher.charAt(i) - (key.charAt(i) - 'a'))
                .map(Cipher::normalize)
                .collect(StringBuilder::new, StringBuilder::appendCodePoint, StringBuilder::append)
                    .toString();
    }
}

File: scala/simple-cipher/src/main/scala/Cipher.scala
import scala.util.Random

object Cipher {
  def apply(optKey: Option[String]) = new Cipher(optKey)
}

class Cipher(optKey: Option[String]) {
  import Cipher._

  for(k <- optKey) {
    require(!k.isEmpty)
    require(k.forall(_.isLower))
  }

  def randomKey(n: Int) = Option(Random.alphanumeric.filter(_.isLower).take(n).mkString)

  lazy val key: String = optKey.orElse(randomKey(26)).get

  def encode(plain: String) =
    mapWithKey(plain, { case (p, k) => (p + (k - 'a')).toChar })

  def decode(cipher: String) =
    mapWithKey(cipher, { case (p, k) => (p - (k - 'a')).toChar })

  def mapWithKey(text: String, f: ((Char, Char)) => Char) =
    (text zip key)
      .map(f)
      .map(normalize)
      .mkString

  def normalize(c: Char) =
    if(c.isLower) c
    else if( c < 'a') (c + 26).toChar
    else (c - 26).toChar
}
